File: ClaseAnimal.cpp
#include <iostream> 

using namespace std;

class Animal{
       public:
       virtual void hablar()=0;
};

class Perro : public Animal{
       public: 
              virtual void hablar(){
                     cout<<"Guau guau";
              }
};

class Gato : public Animal{
       public: 
              virtual void hablar(){
                     cout<<"Miau Miau";
              }
};

int main(){
       Animal A[] ={ new Perro(), new Gato (), new Perro()};
       for (int i = 0; i <3; i++)
              A[i].hablar();
}

File: ClaseAnimal.py
from abc import ABC, abstractmethod
class Animal(ABC):
       @abstractmethod
       def hablar(self):
              pass
       
class Perro (Animal):
       def hablar(self):
              print("Guau")

class Gato (Animal):
       def hablar(self):
              print("Miau")
              
#Script

a = [Perro(), Gato(), Perro()]

for s in a:
       s.hablar() #La s toma los valores de cada una de las posiciones por eso es s